### ISSUE TITLE
Follow up to #4 - MARIDA class aggregation

### ISSUE THREAD

**Issue by gabrieltseng** [antofuller/CROMA#5]
Hi again,

I wasn't sure whether it makes sense to move this to a new issue - I can close this and move the discussion back to #4.

I have a follow up question regarding the MARIDA experiments:

1. The MARIDA paper says that:
> Due to the moderate size of MARIDA and aiming at a Marine Debris-oriented dataset, the initial 15 classes were aggregated to 11 classes. The categories of Wakes, CloudS, Waves and MixWater were grouped with MWater and formed a water super-class, as they semantically belong to the same class as well as present similar spectral profiles (see online material).

Do you follow the original dataset in doing this, or do you predict all 15 classes separately?

**Comment by antofuller** [comment]
Hey Gabriel, no worries.

I used 11 classes, like the original dataset. Also, I just dug up my MARIDA preprocessing code. I combined the validation and test sets, just like Canadian Cropland. I split every 256x256px image into 9 overlapping images of 96x96px (since this is SatMAE's default image size) and discarded images without any labels—like this:

```
def split_into_nine(_img, _label):
    _imgs = []
    _labels = []
    starting_points = [0, 80, 160]
    for x_start in starting_points:
        for y_start in starting_points:
            x_finish = x_start + 96
            y_finish = y_start + 96
            c_img = _img[:, :, x_start:x_finish, y_start:y_finish]
            c_label = _label[:, x_start:x_finish, y_start:y_finish]

            if c_label.unique().shape[0] > 1:
                if not torch.isnan(c_img).any():
                    if not torch.isinf(c_img).any():
                        _imgs.append(c_img)
                        _labels.append(c_label)

I will do my best to upload all my preprocessed datasets this weekend.

**Comment by gabrieltseng** [comment]
Thanks - much appreciated 🙂

**Comment by antofuller** [comment]
> Thanks - much appreciated 🙂

Hi Gabriel, sorry I was mistaken about the number of classes in MARIDA. I used all 15 classes listed in the appendix: *marine debris, dense Sargassum, sparse Sargassum, natural organic material, ship, clouds, marine water, sediment-laden water, foam, turbid water, shallow water, waves, cloud shadows, wakes, and mixed water.*

I've also uploaded my preprocessed versions of the DFC2020, DW-Expert, Canadian Cropland, and MARIDA datasets:
https://huggingface.co/datasets/antofuller/CROMA_benchmarks

Please be sure to cite the original papers if you use them. I share them here to help researchers build on CROMA :)

**Comment by gabrieltseng** [comment]
Thank you!

Since the tiles are overlapping, how did you handle overlapping pixels when computing evaluation set metrics?

Is it possible to use the evaluation code you used to compute IoU ?

**Comment by antofuller** [comment]
Hi Gabriel, I'm very sorry for the 2-week delay.

You can use this code to compute mIoU by providing labels and predictions. They should both be a list of tensors or numpy arrays with elements of shape (N, image_height, image_width). I just looped through all batches in the validation set putting labels and preds into the lists. (this code below was generated by ChatGPT but is mathematically equivalent to the uglier code I actually used :) )

```
def one_hot_encoding(tensor, num_classes):
    # Convert to one-hot encoding
    one_hot = torch.nn.functional.one_hot(tensor, num_classes).float()
    # Rearrange if necessary to move class dimension to the end
    return one_hot

def calculate_miou(labels, preds, num_classes):
    total_intersection = torch.zeros(num_classes)
    total_union = torch.zeros(num_classes)

    for label, pred in zip(labels, preds):
        # Convert NumPy arrays back to PyTorch tensors
        label_tensor = torch.tensor(label)
        pred_tensor = torch.tensor(pred)

        # Convert to one-hot encoding
        one_hot_label = one_hot_encoding(label_tensor, num_classes)
        one_hot_pred = one_hot_encoding(pred_tensor, num_classes)

        # Calculate intersection and union
        intersection = torch.min(one_hot_label, one_hot_pred).sum(dim=0)
        union = torch.max(one_hot_label, one_hot_pred).sum(dim=0)

        # Update totals
        total_intersection += intersection
        total_union += union

    # Calculate IoU for each class and handle division by zero
    iou = total_intersection / total_union.clamp(min=1e-6)

    # Calculate mean IoU
    miou = iou.mean()

    return miou.item()
